functions/src/index.ts

```diff
@@ -242,13 +328,52 @@ exports.calculateAqi = functions.pubsub
           nowCastPm25: nowcastPm25.reading,
           aqi: aqi,
         };
+        aqiBufferData = {
+          aqi: aqi,
+          timestamp: FieldValue.serverTimestamp(),
+        };
       }
-    }
+      // Send AQI reading to current-readings to be displayed on the map
+      await firestore.collection('current-reading').doc('pm25').set({
+        lastUpdated: FieldValue.serverTimestamp(),
+        data: currentData,
+      });
 
-    await firestore.collection('current-reading').doc('pm25').set({
-      lastUpdated: FieldValue.serverTimestamp(),
-      data: currentData,
-    });
+      /// / -------- Update the AQI Circular Buffer -------- ////
+
+      // Document reference for current sensor
+      const sensorDocRef = firestore.collection('/sensors').doc(docId);
+
+      await sensorDocRef.get().then(doc => {
+        if (doc.exists) {
+          const status = doc.get('aqiBufferStatus');
+          if (status === bufferStatus.Exists) {
+            // The buffer exists, proceed with normal update
+            let aqiBufferIndex = doc.get('aqiBufferIndex');
+            const aqiBuffer = doc.get('aqiBuffer');
+
+            // Update circular buffer
+            aqiBuffer[aqiBufferIndex] = aqiBufferData;
+            // Increment buffer index
+            aqiBufferIndex = incrementIndex(aqiBufferIndex, aqiBuffer.length);
+            // Update document
+            sensorDocRef.update({
+              aqiBufferIndex: aqiBufferIndex,
+              aqiBuffer: aqiBuffer,
+            });
+          } else if (
```

Comment:
Like above, should this be an `else`?

Reply:
Like above, no because we do not want to re-initialize the new buffer if it's already started being initialized.